File: week4_tutorial/threads/threads.go
package main

import (
	"fmt"
	"sync"
)

func printNumbers(wg *sync.WaitGroup) {
	defer wg.Done()
	for i := 1; i <= 5; i++ {
		fmt.Println("Number:", i)
	}
}

func printLetters(wg *sync.WaitGroup) {
	defer wg.Done()
	letters := "abcde"
	for _, letter := range letters {
		fmt.Println("Letter:", string(letter))
	}
}

func main() {
	var wg sync.WaitGroup

	wg.Add(2)

	go printNumbers(&wg)
	go printLetters(&wg)

	wg.Wait()

	fmt.Println("Both goroutines have finished")
}

File: week4_tutorial/threads/threads.rs
use std::thread;

fn print_numbers() {
    for i in 1..=5 {
        println!("Number: {}", i);
    }
}

fn print_letters() {
    for letter in "abcde".chars() {
        println!("Letter: {}", letter);
    }
}

fn main() {
    let thread1 = thread::spawn(print_numbers);
    let thread2 = thread::spawn(print_letters);

    thread1.join().expect("Thread 1 panicked");
    thread2.join().expect("Thread 2 panicked");

    println!("Both threads have finished");
}
